presto-main/src/main/java/com/facebook/presto/event/query/QueryMonitor.java

```diff
@@ -192,8 +204,24 @@ public void queryCompletedEvent(QueryInfo queryInfo)
             }
 
             Optional<String> plan = Optional.empty();
-            if (queryInfo.getPlan().isPresent()) {
-                plan = Optional.of(objectMapper.writeValueAsString(queryInfo.getPlan().get()));
+            try {
+                if (queryInfo.getOutputStage().isPresent()) {
+                    // Stats and costs are suppress, since transaction is already completed
+                    plan = Optional.of(textDistributedPlan(
+                            queryInfo.getOutputStage().get(),
+                            functionRegistry,
+                            (node, sourceStats, lookup, session, types) -> UNKNOWN_STATS,
+                            (node, stats, lookup, session, types) -> UNKNOWN_COST,
+                            queryInfo.getSession().toSession(sessionPropertyManager),
+                            false));
+                }
+                if (!plan.isPresent() && queryInfo.getPlan().isPresent()) {
+                    plan = Optional.of(objectMapper.writeValueAsString(queryInfo.getPlan().get()));
+                }
+            }
+            catch (Exception e) {
```

Comment:
why not to fail?

Reply:
Don't want to not fire the event if we can't print a plan.  BTW the create event does the same thing.